File: python/src/bsb/schema_events.py
from __future__ import annotations

from typing import Any, Literal, TypedDict

from .schema import AnyValiDocument, Schema


EventType = Literal["fire-and-forget", "returnable", "broadcast"]
EventCategory = Literal[
    "emitEvents",
    "onEvents",
    "emitReturnableEvents",
    "onReturnableEvents",
    "emitBroadcast",
    "onBroadcast",
]


class FireAndForgetEventSchema(TypedDict, total=False):
    input: Schema
    description: str
    __brand: Literal["fire-and-forget"]


class ReturnableEventSchema(TypedDict, total=False):
    input: Schema
    output: Schema
    description: str
    default_timeout: float
    __brand: Literal["returnable"]


class BroadcastEventSchema(TypedDict, total=False):
    input: Schema
    description: str
    __brand: Literal["broadcast"]


AnyEventSchema = FireAndForgetEventSchema | ReturnableEventSchema | BroadcastEventSchema
CategorySchemaMap = dict[str, AnyEventSchema]
BSBEventSchemas = dict[str, CategorySchemaMap]


class EventExportDefinition(TypedDict, total=False):
    type: EventType
    category: EventCategory
    description: str
    defaultTimeout: float
    inputSchema: AnyValiDocument
    outputSchema: AnyValiDocument | None


class EventSchemaExport(TypedDict, total=False):
    pluginName: str
    version: str
    events: dict[str, EventExportDefinition]
    dependencies: list[dict[str, str]]
    capabilities: dict[str, Any]
    configSchema: dict[str, Any]


def create_fire_and_forget_event(input_schema: Schema, description: str | None = None) -> FireAndForgetEventSchema:
    event: FireAndForgetEventSchema = {"input": input_schema, "__brand": "fire-and-forget"}
    if description:
        event["description"] = description
    return event


def create_returnable_event(
    input_schema: Schema,
    output_schema: Schema,
    description: str | None = None,
    default_timeout: float | None = None,
) -> ReturnableEventSchema:
    event: ReturnableEventSchema = {
        "input": input_schema,
        "output": output_schema,
        "__brand": "returnable",
    }
    if description:
        event["description"] = description
    if default_timeout is not None:
        event["default_timeout"] = default_timeout
    return event


def create_broadcast_event(input_schema: Schema, description: str | None = None) -> BroadcastEventSchema:
    event: BroadcastEventSchema = {"input": input_schema, "__brand": "broadcast"}
    if description:
        event["description"] = description
    return event


def create_event_schemas(schemas: BSBEventSchemas) -> BSBEventSchemas:
    all_names: set[str] = set()
    duplicates: set[str] = set()
    for category in (
        "emitEvents",
        "onEvents",
        "emitReturnableEvents",
        "onReturnableEvents",
        "emitBroadcast",
        "onBroadcast",
    ):
        for name in schemas.get(category, {}).keys():
            if name in all_names:
                duplicates.add(name)
            all_names.add(name)

    if duplicates:
        duplicate_names = ", ".join(sorted(duplicates))
        print(
            "[BSB Warning] Duplicate event names detected: "
            f"{duplicate_names}. Consider unique names across categories."
        )

    return schemas


def get_event_definition(schemas: BSBEventSchemas | None, category: EventCategory, event_name: str) -> AnyEventSchema | None:
    if not schemas:
        return None
    return schemas.get(category, {}).get(event_name)


def export_event_schemas(plugin_name: str, version: str, schemas: BSBEventSchemas | None) -> EventSchemaExport:
    events: dict[str, EventExportDefinition] = {}
    if schemas:
        for category in (
            "emitEvents",
            "onEvents",
            "emitReturnableEvents",
            "onReturnableEvents",
            "emitBroadcast",
            "onBroadcast",
        ):
            for event_name, event_def in schemas.get(category, {}).items():
                output_schema = None
                event_type: EventType = event_def["__brand"]
                if event_type == "returnable":
                    output_schema = event_def["output"].export("extended")

                export_def: EventExportDefinition = {
                    "type": event_type,
                    "category": category,
                    "inputSchema": event_def["input"].export("extended"),
                    "outputSchema": output_schema,
                }
                if "description" in event_def:
                    export_def["description"] = event_def["description"]
                if event_type == "returnable" and "default_timeout" in event_def:
                    export_def["defaultTimeout"] = event_def["default_timeout"]

                events[event_name] = export_def

    return {
        "pluginName": plugin_name,
        "version": version,
        "events": events,
    }


__all__ = [
    "AnyEventSchema",
    "BSBEventSchemas",
    "BroadcastEventSchema",
    "CategorySchemaMap",
    "EventCategory",
    "EventExportDefinition",
    "EventSchemaExport",
    "EventType",
    "FireAndForgetEventSchema",
    "ReturnableEventSchema",
    "create_broadcast_event",
    "create_event_schemas",
    "create_fire_and_forget_event",
    "create_returnable_event",
    "export_event_schemas",
    "get_event_definition",
]

File: go/bsb/schema_events.go
package bsb

import (
	"fmt"

	av "github.com/BetterCorp/AnyVali/sdk/go"
)

// EventBrand distinguishes event schema types at runtime.
type EventBrand string

const (
	BrandFireAndForget EventBrand = "fire-and-forget"
	BrandReturnable    EventBrand = "returnable"
	BrandBroadcast     EventBrand = "broadcast"
)

// FireAndForgetEventSchema defines a fire-and-forget event.
type FireAndForgetEventSchema struct {
	Input       av.Schema
	Description string
	Brand       EventBrand
}

// ReturnableEventSchema defines a request/response event.
type ReturnableEventSchema struct {
	Input          av.Schema
	Output         av.Schema
	Description    string
	DefaultTimeout float64 // seconds
	Brand          EventBrand
}

// BroadcastEventSchema defines a broadcast event.
type BroadcastEventSchema struct {
	Input       av.Schema
	Description string
	Brand       EventBrand
}

// CreateFireAndForgetEvent creates a fire-and-forget event schema.
func CreateFireAndForgetEvent(input av.Schema, description ...string) FireAndForgetEventSchema {
	desc := ""
	if len(description) > 0 {
		desc = description[0]
	}
	return FireAndForgetEventSchema{
		Input:       input,
		Description: desc,
		Brand:       BrandFireAndForget,
	}
}

// CreateReturnableEvent creates a returnable (request/response) event schema.
func CreateReturnableEvent(input, output av.Schema, description string, defaultTimeout ...float64) ReturnableEventSchema {
	timeout := 5.0
	if len(defaultTimeout) > 0 {
		timeout = defaultTimeout[0]
	}
	return ReturnableEventSchema{
		Input:          input,
		Output:         output,
		Description:    description,
		DefaultTimeout: timeout,
		Brand:          BrandReturnable,
	}
}

// CreateBroadcastEvent creates a broadcast event schema.
func CreateBroadcastEvent(input av.Schema, description ...string) BroadcastEventSchema {
	desc := ""
	if len(description) > 0 {
		desc = description[0]
	}
	return BroadcastEventSchema{
		Input:       input,
		Description: desc,
		Brand:       BrandBroadcast,
	}
}

// BSBEventSchemas is the complete event schema container for a plugin.
type BSBEventSchemas struct {
	EmitEvents           map[string]FireAndForgetEventSchema
	OnEvents             map[string]FireAndForgetEventSchema
	EmitReturnableEvents map[string]ReturnableEventSchema
	OnReturnableEvents   map[string]ReturnableEventSchema
	EmitBroadcast        map[string]BroadcastEventSchema
	OnBroadcast          map[string]BroadcastEventSchema
}

// NewEventSchemas creates a new empty BSBEventSchemas.
func NewEventSchemas() BSBEventSchemas {
	return BSBEventSchemas{
		EmitEvents:           make(map[string]FireAndForgetEventSchema),
		OnEvents:             make(map[string]FireAndForgetEventSchema),
		EmitReturnableEvents: make(map[string]ReturnableEventSchema),
		OnReturnableEvents:   make(map[string]ReturnableEventSchema),
		EmitBroadcast:        make(map[string]BroadcastEventSchema),
		OnBroadcast:          make(map[string]BroadcastEventSchema),
	}
}

// Validate checks for duplicate event names across categories.
func (s BSBEventSchemas) Validate() error {
	seen := make(map[string]string) // event name -> category
	check := func(name, category string) error {
		if existing, ok := seen[name]; ok {
			return fmt.Errorf("duplicate event name %q found in both %s and %s", name, existing, category)
		}
		seen[name] = category
		return nil
	}
	for name := range s.EmitEvents {
		if err := check(name, "emitEvents"); err != nil {
			return err
		}
	}
	for name := range s.OnEvents {
		if err := check(name, "onEvents"); err != nil {
			return err
		}
	}
	for name := range s.EmitReturnableEvents {
		if err := check(name, "emitReturnableEvents"); err != nil {
			return err
		}
	}
	for name := range s.OnReturnableEvents {
		if err := check(name, "onReturnableEvents"); err != nil {
			return err
		}
	}
	for name := range s.EmitBroadcast {
		if err := check(name, "emitBroadcast"); err != nil {
			return err
		}
	}
	for name := range s.OnBroadcast {
		if err := check(name, "onBroadcast"); err != nil {
			return err
		}
	}
	return nil
}

// EventSchemaExport represents the exported form of event schemas for
// cross-language client generation.
type EventSchemaExport struct {
	PluginName string                    `json:"pluginName"`
	Version    string                    `json:"version"`
	Events     map[string]EventExportDef `json:"events"`
}

// EventExportDef is a single exported event definition.
type EventExportDef struct {
	Category    string    `json:"category"`
	Description string    `json:"description,omitempty"`
	Input       av.Schema `json:"input,omitempty"`
	Output      av.Schema `json:"output,omitempty"`
	Timeout     float64   `json:"timeout,omitempty"`
}

// ExportSchemas exports event schemas as a portable structure for code generation.
func ExportSchemas(pluginName, version string, schemas BSBEventSchemas) EventSchemaExport {
	export := EventSchemaExport{
		PluginName: pluginName,
		Version:    version,
		Events:     make(map[string]EventExportDef),
	}
	for name, s := range schemas.EmitEvents {
		export.Events[name] = EventExportDef{Category: "emitEvent", Description: s.Description, Input: s.Input}
	}
	for name, s := range schemas.OnEvents {
		export.Events[name] = EventExportDef{Category: "onEvent", Description: s.Description, Input: s.Input}
	}
	for name, s := range schemas.EmitReturnableEvents {
		export.Events[name] = EventExportDef{Category: "emitReturnableEvent", Description: s.Description, Input: s.Input, Output: s.Output, Timeout: s.DefaultTimeout}
	}
	for name, s := range schemas.OnReturnableEvents {
		export.Events[name] = EventExportDef{Category: "onReturnableEvent", Description: s.Description, Input: s.Input, Output: s.Output, Timeout: s.DefaultTimeout}
	}
	for name, s := range schemas.EmitBroadcast {
		export.Events[name] = EventExportDef{Category: "emitBroadcast", Description: s.Description, Input: s.Input}
	}
	for name, s := range schemas.OnBroadcast {
		export.Events[name] = EventExportDef{Category: "onBroadcast", Description: s.Description, Input: s.Input}
	}
	return export
}
